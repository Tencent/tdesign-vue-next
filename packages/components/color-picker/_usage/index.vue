<!-- 该脚本为自动生成，如有需要请在 /script/generate-usage/index.js 中调整 -->
<template>
  <base-usage :code="usageCode" :config-list="configList" :panel-list="panelList" @panel-change="onPanelChange">
    <template #colorPicker="{ configProps }">
      <t-color-picker v-bind="configProps" :default-value="defaultColor" />
    </template>
    <template #colorPickerPanel="{ configProps }">
      <t-color-picker-panel v-bind="configProps" :default-value="defaultColor" />
    </template>
  </base-usage>
</template>

<script setup lang="jsx">
/* eslint-disable */
import { ref } from 'vue';
import colorPickerConfigJson from './props.json';
import colorPickerPanelConfigJson from './panel-props.json';

const configList = ref(colorPickerConfigJson);
const panelList = [
  { label: 'colorPicker', value: 'colorPicker', config: colorPickerConfigJson },
  { label: 'colorPickerPanel', value: 'colorPickerPanel', config: colorPickerPanelConfigJson },
];
const defaultColor = ref('rgb(0, 82, 217)');

const usageCodeMap = {
  colorPicker: '<t-color-picker v-bind="configProps" />',
  colorPickerPanel: '<t-color-picker-panel v-bind="configProps" />',
};
const usageCode = ref(`<template>${usageCodeMap[panelList[0].value].trim()}</template>`);

function onPanelChange(panel) {
  configList.value = panelList.find((item) => item.value === panel).config;
  usageCode.value = `<template>${usageCodeMap[panel].trim()}</template>`;
}
</script>
