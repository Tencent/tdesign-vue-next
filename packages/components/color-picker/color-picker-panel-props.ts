/* eslint-disable */

/**
 * 该文件为脚本自动生成文件，请勿随意修改。如需修改请联系 PMC
 * */

import { TdColorPickerPanelProps } from './type';
import { PropType } from 'vue';

export default {
  /** 颜色模式选择。同时支持单色和渐变两种模式，可仅使用单色或者渐变其中一种模式，也可以同时使用。`monochrome` 表示单色，`linear-gradient` 表示渐变色 */
  colorModes: {
    type: Array as PropType<TdColorPickerPanelProps['colorModes']>,
    default: (): TdColorPickerPanelProps['colorModes'] => ['monochrome', 'linear-gradient'],
  },
  /** 是否禁用组件 */
  disabled: Boolean,
  /** 是否开启透明通道 */
  enableAlpha: Boolean,
  /** 是否允许开启通过点击渐变轴增加渐变梯度，默认开启，关闭时只会存在起始和结束两个颜色 */
  enableMultipleGradient: {
    type: Boolean,
    default: true,
  },
  /** 格式化色值。`enableAlpha` 为真时，`HEX8/RGBA/HSLA/HSVA` 有效 */
  format: {
    type: String as PropType<TdColorPickerPanelProps['format']>,
    default: 'RGB' as TdColorPickerPanelProps['format'],
    validator(val: TdColorPickerPanelProps['format']): boolean {
      if (!val) return true;
      return ['HEX', 'HEX8', 'RGB', 'RGBA', 'HSL', 'HSLA', 'HSV', 'HSVA', 'CMYK', 'CSS'].includes(val);
    },
  },
  /** 最近使用的颜色。值为 [] 表示以组件内部的“最近使用颜色”为准，值长度大于 0 则以该值为准显示“最近使用颜色”。值为 false 或 null 则完全不显示“最近使用颜色” */
  recentColors: {
    type: Array as PropType<TdColorPickerPanelProps['recentColors']>,
    default: undefined as TdColorPickerPanelProps['recentColors'],
  },
  /** 最近使用的颜色。值为 [] 表示以组件内部的“最近使用颜色”为准，值长度大于 0 则以该值为准显示“最近使用颜色”。值为 false 或 null 则完全不显示“最近使用颜色”，非受控属性 */
  defaultRecentColors: {
    type: Array as PropType<TdColorPickerPanelProps['defaultRecentColors']>,
    default: (): TdColorPickerPanelProps['defaultRecentColors'] => [],
  },
  /** 透传 SelectInput 筛选器输入框组件的全部属性 */
  selectInputProps: {
    type: Object as PropType<TdColorPickerPanelProps['selectInputProps']>,
  },
  /** 是否展示颜色选择条右侧的颜色预览区域 */
  showPrimaryColorPreview: {
    type: Boolean,
    default: true,
  },
  /** 系统预设的颜色样例，值为 `null` 或 `[]` 则不显示系统色，值为 `undefined` 会显示组件内置的系统默认色 */
  swatchColors: {
    type: Array as PropType<TdColorPickerPanelProps['swatchColors']>,
  },
  /** 色值 */
  value: {
    type: String,
    default: undefined,
  },
  modelValue: {
    type: String,
    default: undefined,
  },
  /** 色值，非受控属性 */
  defaultValue: {
    type: String,
    default: '',
  },
  /** 选中的色值发生变化时触发，第一个参数 `value` 表示新色值，`context.color` 表示当前调色板控制器的色值，`context.trigger` 表示触发颜色变化的来源 */
  onChange: Function as PropType<TdColorPickerPanelProps['onChange']>,
  /** 调色板控制器的值变化时触发，`context.color` 指调色板控制器的值 */
  onPaletteBarChange: Function as PropType<TdColorPickerPanelProps['onPaletteBarChange']>,
  /** 最近使用颜色发生变化时触发 */
  onRecentColorsChange: Function as PropType<TdColorPickerPanelProps['onRecentColorsChange']>,
};
