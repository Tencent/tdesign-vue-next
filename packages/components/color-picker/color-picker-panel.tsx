import { computed, defineComponent } from 'vue';
import props from './color-picker-panel-props';
import ColorPanel from './components/panel';
import { usePrefixClass } from '@tdesign/hooks';
import { pickBy } from 'lodash-es';

export default defineComponent({
  name: 'TColorPickerPanel',
  inheritAttrs: false,
  props,
  setup(props, { attrs }) {
    const newProps = computed(() => pickBy({ ...props, ...attrs }, (v) => v !== undefined));
    const prefix = usePrefixClass();
    return () => <ColorPanel {...newProps.value} popupProps={null} class={`${prefix.value}-is-inline`} />;
  },
});
