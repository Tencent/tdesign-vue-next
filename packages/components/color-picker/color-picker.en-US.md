:: BASE_DOC ::

## API

### ColorPicker Props

name | type | default | description | required
-- | -- | -- | -- | --
borderless | Boolean | false | \- | N
clearable | Boolean | false | \- | N
colorModes | Array | ["monochrome", "linear-gradient"] | Typescript：`Array<'monochrome' \| 'linear-gradient'>` | N
disabled | Boolean | undefined | \- | N
enableAlpha | Boolean | false | \- | N
enableMultipleGradient | Boolean | true | \- | N
format | String | RGB | When `enableAlpha` is true, `HEX8/RGBA/HSLA/HSVA` are valid。options: HEX/HEX8/RGB/RGBA/HSL/HSLA/HSV/HSVA/CMYK/CSS | N
inputProps | Object | - | Typescript：`InputProps`，[Input API Documents](./input?tab=api)。[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts) | N
popupProps | Object | - | Typescript：`PopupProps`，[Popup API Documents](./popup?tab=api)。[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts) | N
recentColors | Array | [] | used color recently。`v-model:recentColors` is supported。Typescript：`Array<string> \| boolean \| null` | N
defaultRecentColors | Array | [] | used color recently。uncontrolled property。Typescript：`Array<string> \| boolean \| null` | N
selectInputProps | Object | - | Typescript：`SelectInputProps`，[SelectInput API Documents](./select-input?tab=api)。[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts) | N
showPrimaryColorPreview | Boolean | true | \- | N
size | String | medium | options: small/medium/large。Typescript：`SizeEnum`。[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/common.ts) | N
swatchColors | Array | undefined | swatch colors。Typescript：`Array<string> \| null \| undefined` | N
value | String | - | color value。`v-model` and `v-model:value` is supported | N
defaultValue | String | - | color value。uncontrolled property | N
onChange | Function |  | Typescript：`(value: string, context: { color: ColorObject; trigger: ColorPickerChangeTrigger }) => void`<br/>[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts)。<br/>`type ColorPickerChangeTrigger = 'palette-saturation-brightness' \| 'palette-saturation' \| 'palette-brightness' \| 'palette-hue-bar' \| 'palette-alpha-bar' \| 'input' \| 'preset' \| 'recent' `<br/> | N
onClear | Function |  | Typescript：`(context: { e: MouseEvent }) => void`<br/> | N
onPaletteBarChange | Function |  | Typescript：`(context: { color: ColorObject }) => void`<br/>[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts)。<br/>`interface ColorObject { alpha: number; css: string; hex: string; hex8: string; hsl: string; hsla: string; hsv: string; hsva: string; rgb: string; rgba: string; saturation: number; value: number; isGradient: boolean; linearGradient?: string; }`<br/> | N
onRecentColorsChange | Function |  | Typescript：`(value: Array<string>) => void`<br/> | N

### ColorPicker Events

name | params | description
-- | -- | --
change | `(value: string, context: { color: ColorObject; trigger: ColorPickerChangeTrigger })` | [see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts)。<br/>`type ColorPickerChangeTrigger = 'palette-saturation-brightness' \| 'palette-saturation' \| 'palette-brightness' \| 'palette-hue-bar' \| 'palette-alpha-bar' \| 'input' \| 'preset' \| 'recent' `<br/>
clear | `(context: { e: MouseEvent })` | \-
palette-bar-change | `(context: { color: ColorObject })` | [see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker/type.ts)。<br/>`interface ColorObject { alpha: number; css: string; hex: string; hex8: string; hsl: string; hsla: string; hsv: string; hsva: string; rgb: string; rgba: string; saturation: number; value: number; isGradient: boolean; linearGradient?: string; }`<br/>
recent-colors-change | `(value: Array<string>)` | \-


### ColorPickerPanel Props

name | type | default | description | required
-- | -- | -- | -- | --
colorModes | Array | ["monochrome", "linear-gradient"] | Typescript：`Array<'monochrome' \| 'linear-gradient'>` | N
disabled | Boolean | - | \- | N
enableAlpha | Boolean | false | \- | N
enableMultipleGradient | Boolean | true | \- | N
format | String | RGB | When `enableAlpha` is true, `HEX8/RGBA/HSLA/HSVA` are valid。options: HEX/HEX8/RGB/RGBA/HSL/HSLA/HSV/HSVA/CMYK/CSS | N
recentColors | Array | [] | used color recently。`v-model:recentColors` is supported。Typescript：`Array<string> \| boolean \| null` | N
defaultRecentColors | Array | [] | used color recently。uncontrolled property。Typescript：`Array<string> \| boolean \| null` | N
showPrimaryColorPreview | Boolean | true | \- | N
swatchColors | Array | - | swatch colors。Typescript：`Array<string> \| null \| undefined` | N
value | String | - | color value。`v-model` and `v-model:value` is supported | N
defaultValue | String | - | color value。uncontrolled property | N
onChange | Function |  | Typescript：`(value: string, context: { color: ColorObject; trigger: ColorPickerChangeTrigger }) => void`<br/>[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker-panel/type.ts)。<br/>`type ColorPickerChangeTrigger = 'palette-saturation-brightness' \| 'palette-saturation' \| 'palette-brightness' \| 'palette-hue-bar' \| 'palette-alpha-bar' \| 'input' \| 'preset' \| 'recent' `<br/> | N
onPaletteBarChange | Function |  | Typescript：`(context: { color: ColorObject }) => void`<br/>[see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker-panel/type.ts)。<br/>`interface ColorObject { alpha: number; css: string; hex: string; hex8: string; hsl: string; hsla: string; hsv: string; hsva: string; rgb: string; rgba: string; saturation: number; value: number; isGradient: boolean; linearGradient?: string; }`<br/> | N
onRecentColorsChange | Function |  | Typescript：`(value: Array<string>) => void`<br/> | N

### ColorPickerPanel Events

name | params | description
-- | -- | --
change | `(value: string, context: { color: ColorObject; trigger: ColorPickerChangeTrigger })` | [see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker-panel/type.ts)。<br/>`type ColorPickerChangeTrigger = 'palette-saturation-brightness' \| 'palette-saturation' \| 'palette-brightness' \| 'palette-hue-bar' \| 'palette-alpha-bar' \| 'input' \| 'preset' \| 'recent' `<br/>
palette-bar-change | `(context: { color: ColorObject })` | [see more ts definition](https://github.com/Tencent/tdesign-vue-next/blob/develop/packages/components/color-picker-panel/type.ts)。<br/>`interface ColorObject { alpha: number; css: string; hex: string; hex8: string; hsl: string; hsla: string; hsv: string; hsva: string; rgb: string; rgba: string; saturation: number; value: number; isGradient: boolean; linearGradient?: string; }`<br/>
recent-colors-change | `(value: Array<string>)` | \-
