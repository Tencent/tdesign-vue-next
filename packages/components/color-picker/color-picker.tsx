import { defineComponent, ref, toRefs } from 'vue';
import useVModel from '../hooks/useVModel';
import { useTNodeDefault } from '../hooks/tnode';
import props from './props';
import { Popup as TPopup } from '../popup';
import ColorPanel from './components/panel';
import DefaultTrigger from './components/trigger';
import { TdColorContext } from './types';
import { useBaseClassName } from './hooks';

export default defineComponent({
  name: 'TColorPicker',
  props,
  setup(props) {
    const baseClassName = useBaseClassName();
    const renderTNodeJSXDefault = useTNodeDefault();
    const visible = ref(false);
    const setVisible = (value: boolean) => (visible.value = value);

    const { value: inputValue, modelValue } = toRefs(props);
    const [innerValue, setInnerValue] = useVModel(inputValue, modelValue, props.defaultValue, props.onChange);

    const refTrigger = ref<HTMLElement>();

    const handleClear = (context: { e: MouseEvent }) => props.onClear?.(context);

    const renderPopupContent = () => {
      if (props.disabled) {
        return null;
      }
      const newProps = { ...props };
      delete newProps.onChange;
      return (
        <ColorPanel
          {...newProps}
          disabled={props.disabled}
          value={innerValue.value}
          togglePopup={setVisible}
          onChange={(value: string, context: TdColorContext) => setInnerValue(value, context)}
        />
      );
    };

    return () => {
      const popProps = {
        placement: 'bottom-left',
        ...((props.popupProps as any) || {}),
        trigger: 'click',
        attach: 'body',
        overlayClassName: [baseClassName.value],
        visible: visible.value,
        overlayInnerStyle: {
          padding: 0,
        },
        onVisibleChange: (
          visible: boolean,
          context: {
            trigger: string;
          },
        ) => {
          if (context.trigger === 'document') {
            setVisible(false);
          }
        },
      };
      return (
        <TPopup {...popProps} content={renderPopupContent}>
          <div class={`${baseClassName.value}__trigger`} onClick={() => setVisible(!visible.value)} ref={refTrigger}>
            {renderTNodeJSXDefault(
              'default',
              <DefaultTrigger
                borderless={props.borderless}
                color={innerValue.value}
                disabled={props.disabled}
                clearable={props.clearable}
                input-props={props.inputProps}
                onTriggerChange={setInnerValue}
                onTriggerClear={handleClear}
                size={props.size}
              />,
            )}
          </div>
        </TPopup>
      );
    };
  },
});
