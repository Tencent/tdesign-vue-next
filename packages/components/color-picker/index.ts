import _ColorPickerPanel from './color-picker-panel';
import _ColorPicker from './color-picker';
import { withInstall } from '@tdesign/shared-utils';
import { TdColorPickerProps } from './type';

import './style';

export * from './type';
export type ColorPickerProps = TdColorPickerProps;
export type ColorPickerPanelProps = TdColorPickerProps;

export const ColorPickerPanel = withInstall(_ColorPickerPanel);
export const ColorPicker = withInstall(_ColorPicker);
