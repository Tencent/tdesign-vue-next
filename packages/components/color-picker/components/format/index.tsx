import { defineComponent, PropType, ref, watch } from 'vue';
import { upperCase } from 'lodash-es';
import { TdColorPickerProps } from '../../type';
import props from '../../props';
import { FORMATS } from '@tdesign/common-js/color-picker/constants';
import { Color } from '../../utils';
import { Select as TSelect, Option as TOption } from '../../../select';
import FormatInputs from './inputs';
import { useBaseClassName } from '../../hooks';
import type { TdSelectInputProps } from '../../../select-input/type';

export default defineComponent({
  name: 'FormatPanel',
  inheritAttrs: false,
  props: {
    ...props,
    color: {
      type: Object as PropType<Color>,
    },
    onModeChange: {
      type: Function,
      default: () => {
        return () => {};
      },
    },
    onInputChange: {
      type: Function,
      default: () => {
        return () => {};
      },
    },
  },
  setup(props) {
    const baseClassName = useBaseClassName();
    const formatModel = ref<TdColorPickerProps['format']>(props.format);
    watch(
      () => [props.format],
      () => (formatModel.value = props.format),
    );

    /**
     * 格式化类型改变触发
     * @param v
     */
    const handleModeChange = (v: TdColorPickerProps['format']) => {
      formatModel.value = v;
      props.onModeChange(v);
    };

    return () => {
      const newProps = {
        ...props,
        format: formatModel.value,
      };
      const selectInputProps = {
        ...((props.selectInputProps as Object) || {}),
      };
      return (
        <div class={`${baseClassName.value}__format`}>
          <div class={`${baseClassName.value}__format--item`}>
            <TSelect
              size="small"
              class={`${baseClassName.value}__format-mode-select`}
              selectInputProps={{ ...selectInputProps }}
              popupProps={{
                overlayClassName: `${baseClassName.value}__select-options`,
                ...(selectInputProps as TdSelectInputProps).popupProps,
              }}
              v-model={formatModel.value}
              onChange={handleModeChange}
            >
              {FORMATS.map((item) => (
                <TOption key={item} value={item} label={upperCase(item)} style={{ fontSize: '12px' }} />
              ))}
            </TSelect>
          </div>
          <div class={`${baseClassName.value}__format--item`}>
            <FormatInputs {...newProps} />
          </div>
        </div>
      );
    };
  },
});
