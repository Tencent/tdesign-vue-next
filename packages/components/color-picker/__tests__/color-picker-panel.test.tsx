import { ColorObject, ColorPickerPanel } from '@tdesign/components/color-picker';
import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';
import { nextTick, ref } from 'vue';

// every component needs four parts: props/events/slots/functions.
describe('ColorPickerPanel', () => {
  describe(':props', () => {
    it(':colorModes', () => {
      const modes = ref(['monochrome']);
      const wrapper = mount(() => <ColorPickerPanel color-modes={modes.value} />);
      // 单模式不生成mode dom
      expect(wrapper.find('.t-color-picker__mode').exists()).toBe(false);
    });

    it(':disabled', () => {
      const disabled = ref(true);
      const wrapper = mount(() => <ColorPickerPanel disabled={disabled.value} />);
      expect(wrapper.classes()).toContain('t-is-disabled');
    });

    it(':enableAlpha', () => {
      const enableAlpha = ref(true);
      const wrapper = mount(() => <ColorPickerPanel enable-alpha={enableAlpha.value} />);
      expect(wrapper.find('.t-color-picker__alpha').exists()).toBe(true);
    });

    it(':enableMultipleGradient', async () => {
      const testColor = ref('linear-gradient(45deg, #4facfe 0%, #00f2fe 100%)');

      const wrapper = mount(
        <ColorPickerPanel v-model={testColor.value} color-modes={['linear-gradient']} enableMultipleGradient={true} />,
      );
      const sliderNode = wrapper.find('.gradient-thumbs').element;

      const clickEvent = new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
        clientX: 10,
        clientY: 0,
      });

      sliderNode.dispatchEvent(clickEvent);
      await nextTick();
      expect(sliderNode.children.length).toBe(3);
    });

    it(':format', () => {
      const color = ref('red');

      const wrapper = mount(() => <ColorPickerPanel v-model={color.value} format={'HEX'} />);

      const input = wrapper.find('.t-input__inner');
      expect(input.exists()).toBeTruthy();
      expect((input.element as HTMLInputElement).value).toBe('HEX');
      wrapper.unmount();
    });

    it(':recentColors', () => {
      const recentColors = ref(['red', 'yellow', 'blue']);
      const wrapper = mount(() => <ColorPickerPanel recent-colors={recentColors.value} />);
      const swatch = wrapper.find('.t-color-picker__swatches');
      const items = swatch.findAll('.t-color-picker__swatches--item');
      expect(items.length).toBe(recentColors.value.length);
    });

    it(':defaultRecentColors', () => {
      const recentColors = ref(['red', 'yellow', 'blue']);
      const wrapper = mount(() => <ColorPickerPanel defaultRecentColors={recentColors.value} />);
      const swatch = wrapper.find('.t-color-picker__swatches');
      const items = swatch.findAll('.t-color-picker__swatches--item');
      expect(items.length).toBe(recentColors.value.length);
    });

    it(':showPrimaryColorPreview', () => {
      const wrapper = mount(() => <ColorPickerPanel show-primary-color-preview />);
      expect(wrapper.find('.t-color-picker__sliders-preview').exists()).toBe(true);
    });

    it(':swatchColors', () => {
      const wrapper = mount(() => <ColorPickerPanel recent-colors={null} swatch-colors={null} />);
      const swatches = wrapper.findAll('.t-color-picker__swatches');
      expect(swatches.length).toBe(0);
    });

    it(':value', async () => {
      const value = '#0052d9';
      const wrapper = mount(() => <ColorPickerPanel v-model={value} format={'HEX'} />);
      const input = wrapper.findAll('.t-input__inner')[1];
      expect((input.element as HTMLInputElement).value).toBe('#0052d9');
    });

    it(':defaultValue', async () => {
      const wrapper = mount(() => <ColorPickerPanel defaultValue="#0052d9" format={'HEX'} />);
      const input = wrapper.findAll('.t-input__inner')[1];
      expect((input.element as HTMLInputElement).value).toBe('#0052d9');
    });
  });

  describe(':events', () => {
    it('change', async () => {
      const value = ref('#0052d9');
      const handleChange = (val: string) => {
        value.value = val;
      };
      const recentColors = ref(['red']);

      const wrapper = mount(() => (
        <ColorPickerPanel recent-colors={recentColors.value} v-model={value.value} onChange={handleChange} />
      ));
      const swatch = wrapper.find('.t-color-picker__swatches');
      const item = swatch.find('.t-color-picker__swatches--item');
      item.trigger('click');
      await nextTick();
      expect(value.value).toBe('rgb(255, 0, 0)');
    });
    // it('palette-bar-change', async () => {
    //   const value = ref('#0052d9');
    //   const result = ref<ColorObject>();

    //   const handlePaletteBarChange = (context: { color: ColorObject }) => {
    //     result.value = context.color;
    //   };

    //   const wrapper = mount(() => (
    //     <ColorPickerPanel v-model={value.value} onPaletteBarChange={handlePaletteBarChange} />
    //   ));
    //   await nextTick();

    //   const slider = wrapper.find('.t-color-picker__rail');

    //   const clickEvent = new MouseEvent('click', {
    //     bubbles: true,
    //     cancelable: true,
    //     clientX: 20,
    //   });

    //   slider.element.dispatchEvent(clickEvent);
    //   await nextTick();
    //   expect(result.value).toMatchSnapshot();
    // });
    it('recent-colors-change', async () => {
      const value = ref('#0052d9');
      const recentColors = ['red'];
      const result = ref<Array<string>>([]);
      const handleRecentColorsChange = (value: Array<string>) => {
        result.value = value;
      };

      const wrapper = mount(() => (
        <ColorPickerPanel
          recent-colors={recentColors}
          v-model={value.value}
          onRecentColorsChange={handleRecentColorsChange}
        />
      ));

      const swatch = wrapper.find('.t-color-picker__swatches');
      const items = swatch.findAll('.t-color-picker__icon');
      items[0].trigger('click');
      await nextTick();

      expect(result.value).toEqual(['rgba(0, 82, 217, 1)', 'red']);
    });
  });
});
