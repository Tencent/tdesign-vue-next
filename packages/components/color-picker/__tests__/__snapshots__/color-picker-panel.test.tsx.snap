// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorPickerPanel > :events > palette-bar-change 1`] = `{
    "alpha": 1,
    "css": "rgba(0, 163, 217, 1)",
    "hex": "#00a3d9",
    "hex8": "#00a3d9ff",
    "hsl": "hsl(195, 100%, 43%)",
    "hsla": "hsla(195, 100%, 43%, 1)",
    "hsv": "hsv(195, 100%, 85%)",
    "hsva": "hsva(195, 100%, 85%, 1)",
    "rgb": "rgb(0, 163, 217)",
    "rgba": "rgba(0, 163, 217, 1)",
    "saturation": 1,
    "value": 0.8509803921568627,
    "isGradient": false
}`;
